### PR TITLE
[ENHANCEMENT] change release method of ByteBuf in decode phrase

### PR DESCRIPTION
**Pull request by wuzhanpeng**
Motivation: In some scenarios, the execution of `writeAndFlush` may be asynchronous, thus it might be better to use `addListener` to accomplish asynchronous callbacks.

### CODE REVIEW

**Comment by BewareMyPower** [comment]
It looks like `KafkaRequestHandlerTest` is stuck, could you take a look?

**Review by jiazhai — APPROVED** [review]

**Comment by BewareMyPower** [comment]
Recently I found an exception about reference count from `ResponseCallbackWrapper` before this PR

```
09:45:02.834 [pulsar-io-20-4:io.streamnative.pulsar.handlers.kop.KafkaCommandDecoder@336] ERROR io.streamnative.pulsar.handlers.kop.KafkaCommandDecoder - Error in executing callback after sending response,
 io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
    at org.apache.bookkeeper.mledger.util.AbstractCASReferenceCounted.release0(AbstractCASReferenceCounted.java:99) ~[managed-ledger-2.8.0-rc-202104202206.jar:2.8.0-rc-202104202206]
    at org.apache.bookkeeper.mledger.util.AbstractCASReferenceCounted.release(AbstractCASReferenceCounted.java:87) ~[managed-ledger-2.8.0-rc-202104202206.jar:2.8.0-rc-202104202206]
    at java.util.ArrayList.forEach(ArrayList.java:1259) ~[?:1.8.0_292]
    at io.streamnative.pulsar.handlers.kop.MessageFetchContext.lambda$null$13(MessageFetchContext.java:405) ~[pulsar-protocol-handler-kafka-2.8.0-SNAPSHOT.jar:?]
    at java.util.concurrent.ConcurrentHashMap$EntrySetView.forEach(ConcurrentHashMap.java:4795) ~[?:1.8.0_292]
    at io.streamnative.pulsar.handlers.kop.MessageFetchContext.lambda$null$14(MessageFetchContext.java:404) ~[pulsar-protocol-handler-kafka-2.8.0-SNAPSHOT.jar:?]
    at org.apache.kafka.common.requests.ResponseCallbackWrapper.responseComplete(ResponseCallbackWrapper.java:46) ~[pulsar-protocol-handler-kafka-2.8.0-SNAPSHOT.jar:?]
    at io.streamnative.pulsar.handlers.kop.KafkaCommandDecoder.writeAndFlushResponseToClient(KafkaCommandDecoder.java:333) ~[pulsar-protocol-handler-kafka-2.8.0-SNAPSHOT.jar:?]
    at io.streamnative.pulsar.handlers.kop.KafkaCommandDecoder.lambda$null$0(KafkaCommandDecoder.java:176) ~[pulsar-protocol-handler-kafka-2.8.0-SNAPSHOT.jar:?]
    at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) [netty-common-4.1.63.Final.jar:4.1.63.Final]
    at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472) [netty-common-4.1.63.Final.jar:4.1.63.Final]
    at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384) [netty-transport-native-epoll-4.1.63.Final-linux-x86_64.jar:4.1.63.Final]
    at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.63.Final.jar:4.1.63.Final]
    at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.63.Final.jar:4.1.63.Final]
    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.63.Final.jar:4.1.63.Final]
    at java.lang.Thread.run(Thread.java:748) [?:1.8.0_292]
```

Do you know why it happened and could this PR avoid it?

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaCommandDecoder.java [hunk 1]
```diff
@@ -309,7 +309,19 @@ protected void writeAndFlushResponseToClient(Channel channel) {
                 ByteBuf result = responseToByteBuf(response.getResponseFuture().get(
                         kafkaConfig.getRequestTimeoutMs(),
                         TimeUnit.MILLISECONDS), response.getRequest());
-                channel.writeAndFlush(result);
+                channel.writeAndFlush(result)
+                        .addListener(future -> {
+                            if (future.isDone()) {
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaCommandDecoder.java:314]: The check of `future.isDone()` is redundant, see  https://netty.io/4.1/api/io/netty/channel/ChannelFuture.html#addListener-io.netty.util.concurrent.GenericFutureListener-

>  The specified listener is notified when this future is done.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

**Comment by BewareMyPower** [comment]
I've opened https://github.com/streamnative/kop/pull/465 to do some refactors on `writeAndFlushResponseToClient` and it contains the change of this PR.

**Pull request closed without merging**